aspnetcore/fundamentals/host/platform-specific-configuration.md

```diff
@@ -0,0 +1,289 @@
+
+The [sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/fundamentals/host/platform-specific-configuration/samples/) includes a Razor Pages app, *HostingStartupApp*, and a class library, *HostingStartupLibrary*. The class library:
+
+* Contains a hosting startup class, `ServiceKeyInjection`, which implements `IHostingStartup`. `ServiceKeyInjection` adds a pair of service strings to the app's configuration using the in-memory configuration provider ([AddInMemoryCollection](/dotnet/api/microsoft.extensions.configuration.memoryconfigurationbuilderextensions.addinmemorycollection)).
+* Includes a `HostingStartup` attribute that identifies the hosting startup's namespace and class.
+
+The `ServiceKeyInjection` class's [Configure](/dotnet/api/microsoft.aspnetcore.hosting.ihostingstartup.configure) method uses an [IWebHostBuilder](/dotnet/api/microsoft.aspnetcore.hosting.iwebhostbuilder) to add enhancements to an app. `IHostingStartup.Configure` in the hosting startup assembly is called by the runtime before `Startup.Configure` in user code, which allows user code to overwrite any configuration provided by the hosting startup assembly.
+
+*HostingStartupLibrary/ServiceKeyInjection.cs*:
+
+[!code-csharp[](platform-specific-configuration/samples/2.x/HostingStartupLibrary/ServiceKeyInjection.cs?name=snippet1)]
+
+The app's Index page reads the configuration values for the two keys set by the class library's hosting startup assembly. The configuration values for these keys are displayed when the page is rendered.
+
+*HostingStartupApp/Pages/Index.cshtml.cs*:
+
+[!code-csharp[](platform-specific-configuration/samples/2.x/HostingStartupApp/Pages/Index.cshtml.cs?name=snippet1&highlight=7-8,13-14)]
+
+The compiled class library is bin-deployed to the app, and the class library's assembly name is listed in the `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES` environment variable. The environment variable is a semicolon-delimited list of assemblies that contain a `HostingStartup` attribute that identifies a hosting startup enhancement to load.
+
+The class library's assembly name can also be provided using the [Hosting Startup Assemblies](xref:fundamentals/host/web-host#hosting-startup-assemblies) host configuration setting.
+
+When multiple hosting startup assembles are present, their [Configure](/dotnet/api/microsoft.aspnetcore.hosting.ihostingstartup.configure) methods are executed in the order that the assemblies are listed.
+
+## Assembly activation
+
+This section describes how to activate a hosting startup from a bin-deployed or runtime store-deployed assembly.
+
+### Create the assembly
+
+An `IHostingStartup` enhancement is deployed as an assembly based on a console app without an entry point. The assembly references the [Microsoft.AspNetCore.Hosting.Abstractions](https://www.nuget.org/packages/Microsoft.AspNetCore.Hosting.Abstractions/) package:
```

Comment:
I think the wrapper project and getting the right deps file came about from `@pakrym`'s comment here :point_right: https://github.com/aspnet/Docs/issues/4438#issuecomment-349041919 (referring to answering me on 4 and 7).

What's the deal with the hosting startup on ...

* Compile-time reference present: No deps file needed (no additional deps specification)
* No compile-time reference: Deps file needed (additional deps specification required)

If you can help me understand that, I think a few words in the topic on it would help readers understand the requirement in the no-compile-time reference scenario.

Reply:
Core will only load assemblies listed in a deps file. When you have a compile time reference then the dependent assembly and all of its own dependencies end up in the app deps file.  When you don't have a compile time reference then you need to provide a supplementary deps file with all of that assemblies dependencies.

Reply:
Under the class lib approach if the class lib's assembly is placed into the runtime store, is an explicit compile-time reference required in the consuming app's project file?

I have it bin-deployed in the text and sample app with a reference in the sample app's project file ..

```xml
<ItemGroup>
  <Reference Include=".\bin\Debug\netcoreapp2.1\HostingStartupLibrary.dll">
    <HintPath>.\bin\Debug\netcoreapp2.1\HostingStartupLibrary.dll</HintPath>
    <SpecificVersion>False</SpecificVersion> 
  </Reference>
</ItemGroup>
```

I'm just wondering in the runtime store case what the compile-time reference looks like. We may need to add it.

Reply:
No, it doesn't require a compile time reference, just an additional deps. That's how we ended up using it for AppInsights and AppServices. Go look at C:\Program Files\dotnet\additionalDeps\Microsoft.AspNetCore.AzureAppServices.HostingStartup.

Reply:
What's the practical difference between:

* A class lib in the runtime store with additional deps
* A console app without an entry point in the runtime store with addtional deps

Reply:
> No, it doesn't require a compile time reference, just an additional deps. 

This has taken on new complexity: The explanation for additional deps under the "assembly activation" approach (console app w/o entry point) now has to be duplicated under class lib with runtime store deployment.

I'm going to have to mull this over a bit. I'll sleep on it and see what I can come up with in the morning.

Reply:
There's a lot of overlapping parts between the scenarios. What if we start by explaining the smaller components and then pull them together in the two main scenarios at the end? And those scenarios build on eachother. E.g. the no-compile-time-reference scenario is an expansion on top of the compile time reference scenario that uses all the same mechanics and adds the additional deps.

Reply:
Yes, that sounds good (devil might be in the details tho). I'll take a good night sleep + up early + plenty of ☕️ and see how it goes in the morning.

Reply:
I need to know this tho -------

What's the practical difference between:

* A class lib in the runtime store with additional deps
* A console app without an entry point in the runtime store with addtional deps

Reply:
@pakrym ?

Reply:
Apparently if the **console app w/o entry point assembly** is bin-deployed with a compile-time ref, there's no need to deal with the deps file (just like in the console lib assembly case).

### Assembly type (delivery options)

* Class lib in a local assembly (runtime store, bin, or NuGet)
* Console app w/o entry point (runtime store, bin, or NuGet - can it be packed?)

### Delivery options

* runtime store without compile-time ref and provide deps
* bin with compile-time ref
* NuGet package with compile-time ref

### Outline

🥁 roll plz ...

[This outline is in the order of presentation.]

* Class lib (only how to create the assembly here)
* Console app w/o entry point (only how to create the assembly here)
* Activation
  - List in `ASPNETCORE_HOSTINGSTARTUPASSEMBLIES`
  - Runtime store
    * Where to place assembly
    * Provide deps instructions
  - NuGet - I still prefer to keep NuGet separate from class lib and bin-deployment. With this outline, there should be minimal duplication of text among the sections.
    * Link to NuGet packing instructions
    * No deps work required
    * Deployment
      - NuGet package via nuget.org (https://docs.microsoft.com/dotnet/core/deploying/creating-nuget-packages and https://docs.microsoft.com/nuget/create-packages/publish-a-package)
      - NuGet package in the runtime store (https://docs.microsoft.com/dotnet/core/deploying/runtime-store)
  - bin with compile-time ref
    * Where to place assembly
    * No deps work required
* Sample app - Unlike the topic today, completely separate the sample description and steps out of the text. Only include in the sample the most likely deployment approaches.

The pieces are just about all here, so I'm not sorry about the work that's been done thus far. However, let's settle on an outline (maybe that one :point_up:) before I continue.

### Posers, questions, queries

1. What's the practical difference between **class lib** and **console app w/o entry point** approaches? Since it looks like they can both delivered with any delivery approach here, it comes down to the practical difference between them. Which one makes sense for given scenarios?
1. Are **class lib+NuGet package** and **console app w/o entry point+runtime store** the most common anticipated assembly+delivery approaches? If so, the sample only needs to show these two.
1. Between **NuGet package on nuget.org** and **NuGet package in the runtime store**, shall we just link out, or shall we link out with guidance on pros+cons?
1. Can we say something about **installer options**? ... making delivery dead-on simple in multi-machine environments without using NuGet? ... or do we just lean in the direction of **_you should always use NuGet!_**